### PR TITLE
Deploy to okfn server for better https support

### PR DESCRIPTION
**Pull request by k-nut**
This pull request aims to solve #4 by deploying to a codefor.de server that we have full control over so that we can make sure that http works.
The goal was not to lose any of the comfort that we currently have with github pages where we do not have to do anything to deploy.

Therefore I created a setup on a virtual machine and a [fabric](http://www.fabfile.org/) deployment script which can be run on travis to automatically deploy the newest version whenever the tests pass.

### On the server
- Nginx serves static files from `/var/www/newest/`
- `/var/www/newest` is symlinked to `/home/deploy/verisons/newest`
- The `/home/deploy/versions` folder contains the `newest` folder and one folder per deployed revision. The newest revision itself is symlinked to `newest` (and through that to `/var/www/newest` which nginx serves).

### With fabric
- We bundle all files in the repository into a `.tar.gz`
- This tarfile is uploaded to the server
- Is is extracted and placed under `/home/deploy/versions/<current_revision>`
- `/home/deploy/versions/<current_revision>` is symlinked to `/home/deploy/versions/newest`
and therefore set to be served by nginx

Doing all the symlink helps us to easily roll back to a former revision if anything ever goes wrong with a deployment. We just need to change where `/home/deploy/versions/newest` points and we are done.

## Problems & last steps needed
One problem here is that by creating an archive for the whole project we also include files like `.git` and all other files that do not have to be deployed. I therefore propose moving the `index.html`, assets, javascript etc. into an `app` folder and just deploying that.

As a last step we need to create a new ssh key for travis and add that to the server.

Todo
- [ ] move files to `app` directory (optional)
- [x] add deploy key to travis
- [x] add travis deploy key to `/home/deploy/.ssh/allowed_hosts`
- [x] change travis script to also deploy
- [x] change CNAME to point to new server

## 🚀 

I hope that I made clear how I see this working. 

I am happy to answer all questions and looking forward to your feedback!

### CODE REVIEW

File: fabfile.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+def deploy():
+    current_revision = local('git rev-parse HEAD', capture=True)
+    create_artifact(current_revision)
+    upload_artifact(current_revision)
+    make_active(current_revision)
+
+def create_artifact(current_revision):
+    """ compress all files into a .tar.gz archive for uploading """
+    archive_path = '/tmp/{revision}.tar.gz'.format(revision=current_revision)
+    local('tar -czf {archive_path} --exclude=.git *'.format(archive_path=archive_path))
```
Comment (torfsen) [inline fabfile.py]: Do we need that `--exclude` if we use a separate `app` folder?

File: fabfile.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
+from fabric.api import run, env, local, put, cd
```
Comment (torfsen) [inline fabfile.py]: I'd love to see a (short) high-level comment on what's being done in the file.

File: fabfile.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    make_active(current_revision)
+
+def create_artifact(current_revision):
+    """ compress all files into a .tar.gz archive for uploading """
+    archive_path = '/tmp/{revision}.tar.gz'.format(revision=current_revision)
+    local('tar -czf {archive_path} --exclude=.git *'.format(archive_path=archive_path))
+
+def upload_artifact(revision):
+    """ upload the archive to the server and extract it """
+    remote_versions_path = '/home/deploy/versions'
```
Comment (torfsen) [inline fabfile.py]: Please move the shared path-prefixes (e.g. `/home/deploy`) into global constants to ease maintenance.

**Comment by johnjohndoe** [comment]
@k-nut Please add some high level documentation to your pull request which describes the deployment. Please state what triggers the deployment, which parts are automated and when it happens.
Please add a "Deployment" section to bottom of the _README.md_. Dependent on the amount of documentation it might make sense to create a separate _DEPLOYMENT.md_ and link to it.

**Comment by torfsen** [comment]
Thanks for the work, @k-nut! I'd prefer a simpler setup, but it seems that you've put some thoughts into making this as simple as possible.

One thing that's missing from my point of view is documentation for this setup and the deployment process as part of the repository itself. You can add it to the README or to a separate document, and it doesn't have be much -- basically what you've wrote in your PR comment in its final form.

By the way, what's the easiest way to test this PR?

File: fabfile.py (hunk 1 above)
Comment (k-nut) [inline fabfile.py]: nope. In that case we could just create a `.tar.gz` for `app` and deploy that

File: fabfile.py (hunk 3 above)
Comment (k-nut) [inline fabfile.py]: Done in 6a2275a

**Comment by k-nut** [comment]
@johnjohndoe @torfsen I updated the readme in 8d1bd35. Is this sufficient for you?

**Comment by k-nut** [comment]
In order for travis to be able to deploy I changed the language to `python` which is needed for fabric. Travis still allows you to use `npm` in that case which is all we need. You can see an example of a build and deploy here: https://travis-ci.org/wo-ist-markt/wo-ist-markt.github.io/builds/149556355.

It looks to me though like one of our functions is failing because we are using the new ecmascript 6 `=>` syntax. I think we have to change this to old javascript if we want to make sure that the test works.

**Comment by k-nut** [comment]
I hope that I addressed all of your points. Let me know if anything is still unclear / needs to be updated. 

@torfsen I think the best way to "test" this is to either see what is happening or travis or to run the deployment locally. For that you would need to give me your public ssh key so that I can add you to the server. That might be a good idea anyways so that I am not the only one who could fix things there.

File: README.md [hunk 4]
```diff
@@ -146,3 +146,16 @@ Tests can be run using [npm][npm]:
 [witten-markets]:http://www.stadtmarketing-witten.de/einkaufen/wochenmaerkte.html
 [wuppertal-wikipedia]:https://en.wikipedia.org/wiki/Wuppertal
 [wuppertal-markets]:https://www.wuppertal.de/tourismus-freizeit/einkaufen/102370100000204430.php
+
+
+## Deployment
```
Comment (johnjohndoe) [inline README.md]: Please consolidate the wording - I recommend using `Travis CI` not `travis` or `Travis`.

File: README.md [hunk 5]
```diff
@@ -146,3 +146,16 @@ Tests can be run using [npm][npm]:
...
+When new code is pushed to master travis runs a test suite against it. If those test
+pass [fabric](fabfile.org) is used to deploy the new version to the server running
+under https://wo-ist-markt.de.
+
+The command that travis executes is `fab deploy -i markt_deploy_id_rsa -H deploy@kiesinger.okfn.de:2207`
+where markt_deploy_id_rsa is an ssh key to log into the server. It is commited to this repository in
```
Comment (johnjohndoe) [inline README.md]: Please format `where markt_deploy_id_rsa` with backticks.

File: fabfile.py (hunk 1 above)
Comment (johnjohndoe) [inline fabfile.py]: I voted for staying with the current folder structure. I think the deployment can just filter out the _.git/_ folder.

File: .travis.yml [hunk 6]
```diff
@@ -1,4 +1,18 @@
...
+
+install:
+  - pip install -r requirements.txt
+  - npm install
+
+script:
+  - npm test
+
+after_success:
+  - test $TRAVIS_PULL_REQUEST == "false" && test $TRAVIS_BRANCH == "master" && fab deploy -i markt_deploy_id_rsa -H deploy@kiesinger.okfn.de:2207
```
Comment (johnjohndoe) [inline .travis.yml]: A short but descriptive comment would be nice here.

File: .travis.yml [hunk 7]
```diff
@@ -1,4 +1,18 @@
-language: node_js
-node_js:
-    - stable
+language: python
 
+python:
+    - '2.7'
+
+before_install:
+    - openssl aes-256-cbc -K $encrypted_02f86fe65d47_key -iv $encrypted_02f86fe65d47_iv
```
Comment (johnjohndoe) [inline .travis.yml]: A short but descriptive comment would be nice here.

File: fabfile.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    run('mkdir -p %s' % destination_path)
+    run('tar xfz %s -C %s' % (source_path, destination_path))
+
+def make_active(revision):
+    """ change the `newest` symlink to point to this revision """
+    run('ln -sfn {base}/{revision}/ {base}/newest'.format(base=BASE_PATH,
+                                                          revision=revision))
+
+def list_versions():
+    run('ls {base}'.format(base=BASE_PATH))
```
Comment (johnjohndoe) [inline fabfile.py]: not `ls -l`?

File: README.md (hunk 4 above)
Comment (k-nut) [inline README.md]: ✅ a61bff0fee3bc0ee4949272ac536fb04138c2060

File: .travis.yml (hunk 7 above)
Comment (k-nut) [inline .travis.yml]: Added comments in bfdaf605e2195c28b378d792fd696a5e796b3f38

File: .travis.yml (hunk 6 above)
Comment (k-nut) [inline .travis.yml]: Added comments in bfdaf605e2195c28b378d792fd696a5e796b3f38

File: fabfile.py (hunk 8 above)
Comment (k-nut) [inline fabfile.py]: ✅  	85acbf1

File: README.md (hunk 5 above)
Comment (k-nut) [inline README.md]: So would you drop that paragraph entirely?

**Comment by k-nut** [comment]
Just pushed another change that replaces the mentioned `=>` fatArrow syntax from the test so that travis can execute them properly. 876c2b2

**Comment by torfsen** [comment]
Some minor issues, but otherwise looks great now!

**Comment by k-nut** [comment]
So @johnjohndoe proposed rewriting this into fewer logical commits. Should I do it or do you think that this is sufficient? And if yes do you have a proposed way for resturcturing? Again, this might be a chance to use a [squash commit](https://github.com/blog/2141-squash-your-commits) maybe?

Or are there any other changes that you would like to see before merging?

**Comment by k-nut** [comment]
I just rebased yesterday. So can we just use the squash commit feature to merge this?

**Comment by johnjohndoe** [comment]
Ah, good. Either, we use the GitHub squash feature or you can do it yourself to edit the commit message as you want it to be. Are there additional changes required with regards to #125?

File: README.md (hunk 5 above)
Comment (johnjohndoe) [inline README.md]: After the latest problems I would leave the paragraph in there. Please format the `where markt_deploy_id_rsa` with backticks. Please replace the last sentence with: The command to deploy manually is: `fab deploy -H deploy@kiesinger.okfn.de:2207`.

**Comment by k-nut** [comment]
🔨 squashed, updated and rebased. I think this is good to 🚢

**Comment by johnjohndoe** [comment]
:rocket: merged! Thanks a lot for this nice piece of work!

**Pull request merged**